### PR TITLE
Vwp/chatregtests

### PR DESCRIPTION
**Pull request by vowelparrot**
This is a draft that starts to refactor the agents as well. I make a draft that avoids that here https://github.com/hwchase17/langchain/pull/3572

### CODE REVIEW

File: langchain/tools/structured.py [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+    inferred_model = validate_arguments(func).model  # type: ignore
+    # Pydantic adds placeholder virtual fields we need to strip
+    filtered_args = get_filtered_args(inferred_model, func)
+    return _create_subset_model(
+        f"{model_name}Schema", inferred_model, list(filtered_args)
+    )
+
+
+class BaseStructuredTool(
+    AbstractBaseTool[Dict, SCHEMA_T, OUTPUT_T], Generic[SCHEMA_T, OUTPUT_T]
```
Comment (vowelparrot) [inline langchain/tools/structured.py:203]: I dislike how many parameters this is taking

File: tests/regression_tests/multi_input_tools.py
Comment (vowelparrot) [inline tests/regression_tests/multi_input_tools.py:1]: ignore this just running some tests

**Comment by vowelparrot** [comment]
I don't like this more than the current implementation.

File: langchain/schema.py [hunk 2]
```diff
@@ -41,10 +41,21 @@ class AgentAction(NamedTuple):
     """Agent's action to take."""
 
     tool: str
-    tool_input: Union[str, dict]
+    tool_input: str
     log: str
 
 
+class StructuredAgentAction(NamedTuple):
```
Comment (eyurtsev) [inline langchain/schema.py:48]: Nit: Any reason to keep on using named tuples instead of data-classes? NamedTuples feel like a python 2 thing

**Comment by vowelparrot** [comment]
I think I prefer https://github.com/hwchase17/langchain/pull/3572 to this one and don't feel an urgent need to refactor all the agents and agent executors

File: langchain/schema.py (hunk 2 above)
Comment (vowelparrot) [reply langchain/schema.py:48]: I'm not sure honestly, but I'm also not sure what data classes would give us apart from a nice to_dict() method?

**Pull request closed without merging**